alerts/send.go

```diff
@@ -0,0 +1,94 @@
+package alerts
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/sideshow/apns2"
+	"github.com/sideshow/apns2/payload"
+	"github.com/sideshow/apns2/token"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/errors"
+)
+
+// Pusher abstracts the underlying push notification mechanism.
+type Pusher interface {
+	// Push a notification to a device.
+	//
+	// deviceToken should uniquely identify a device, while Notification
+	// abstracts the provider-specific notification payload.
+	Push(ctx context.Context, deviceToken devicetokens.DeviceToken, notification *Notification) error
+}
+
+// Notification models a provider-independent push notification.
+type Notification struct {
```

Comment:
Do we have plans for supporting OS-specific push notification data? Are there product requirements regarding the different OS experiences? Have the different OS experiences been considered from a product perspective? Non-technical folks may not understand how push notifications differ between iOS and Android. For example, sounds, numbered badges, message threading, localization, visibility, priority, etc.

A good place to get some ideas is Firebase Cloud Messaging (FCM). FCM supports both iOS and Android push notifications (as well as web). Take a look at https://pkg.go.dev/firebase.google.com/go/messaging#Message. It has a few common fields, but also very specific iOS and Android fields.

I suppose we could bury this OS-specific configuration within the implementation and from the API/interface side only allow Title and Message. I suppose it depends upon whether this solution will be our general PN system or if it will be very specific to the Care Partner app alerts. If more general, then I suspect we'll need to expose more OS-specific options.

Reply:
The short answer is that the mobile team had already decided to use APNS, and are prepared with that workflow.

I've no doubt we'll be supporting FCM at some point, so I made sure to try not to paint myself into any corners. But in the meantime, the spec called for platform to accept APNS tokens and use them to send push notifications.

I read enough about how APNS and FCM both work to be confident that we can easily support both with a minimum of fuss. We'd need to put some thin abstraction layer in anyway, so that we can test the code.

Reply:
As for generality vs specificity... I'm not aware of anything other than the care partner app that will be sending push notifications right now. Therefore my vote is to keep it as simple as possible until we have some idea of what more we'd actually need.